File: services/sidecar/src/workflow_sidecar/protocol.py
"""Versioned NDJSON protocol primitives shared by the sidecar runtime."""

from __future__ import annotations

from dataclasses import dataclass
import json
from typing import Any


PROTOCOL_VERSION = 1
MAX_MESSAGE_BYTES = 1024 * 1024

JsonObject = dict[str, Any]


@dataclass(frozen=True, slots=True)
class Request:
    """A validated host-to-sidecar request."""

    id: str
    method: str
    params: JsonObject


class ProtocolError(ValueError):
    """An error that is safe to expose through the IPC envelope."""

    def __init__(self, code: str, message: str, request_id: str | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.request_id = request_id


def decode_request(payload: bytes) -> Request:
    """Decode and validate one NDJSON request line."""

    if len(payload) > MAX_MESSAGE_BYTES:
        raise ProtocolError("MESSAGE_TOO_LARGE", "Message exceeds the 1 MiB limit")

    try:
        value = json.loads(payload)
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise ProtocolError("INVALID_JSON", "Message is not valid UTF-8 JSON") from exc

    if not isinstance(value, dict):
        raise ProtocolError("INVALID_ENVELOPE", "Message envelope must be an object")

    request_id = value.get("id") if isinstance(value.get("id"), str) else None
    if value.get("v") != PROTOCOL_VERSION:
        raise ProtocolError(
            "UNSUPPORTED_VERSION",
            f"Only protocol version {PROTOCOL_VERSION} is supported",
            request_id,
        )
    if value.get("type") != "request":
        raise ProtocolError(
            "INVALID_ENVELOPE", "Host messages must have type=request", request_id
        )

    method = value.get("method")
    params = value.get("params", {})
    if not request_id or not isinstance(method, str) or not method:
        raise ProtocolError(
            "INVALID_REQUEST", "Request id and method must be non-empty strings", request_id
        )
    if not isinstance(params, dict):
        raise ProtocolError("INVALID_REQUEST", "Request params must be an object", request_id)

    return Request(id=request_id, method=method, params=params)


def success_response(request_id: str, result: JsonObject) -> JsonObject:
    return {
        "v": PROTOCOL_VERSION,
        "type": "response",
        "id": request_id,
        "result": result,
    }


def error_response(
    request_id: str | None,
    code: str,
    message: str,
    diagnostic_id: str | None = None,
) -> JsonObject:
    error: JsonObject = {"code": code, "message": message}
    if diagnostic_id is not None:
        error["diagnostic_id"] = diagnostic_id
    return {
        "v": PROTOCOL_VERSION,
        "type": "response",
        "id": request_id,
        "error": error,
    }


def event(name: str, data: JsonObject) -> JsonObject:
    return {"v": PROTOCOL_VERSION, "type": "event", "event": name, "data": data}

File: apps/desktop/src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

pub const PROTOCOL_VERSION: u32 = 1;
pub const MAX_MESSAGE_BYTES: usize = 1024 * 1024;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct RpcError {
    pub code: String,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub diagnostic_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum InboundMessage {
    Response {
        id: Option<String>,
        result: Option<Value>,
        error: Option<RpcError>,
    },
    Event {
        event: String,
        data: Value,
    },
}

#[derive(Debug, Error)]
pub enum ProtocolDecodeError {
    #[error("message exceeds the 1 MiB limit")]
    MessageTooLarge,
    #[error("invalid JSON: {0}")]
    InvalidJson(#[from] serde_json::Error),
    #[error("unsupported protocol version")]
    UnsupportedVersion,
    #[error("invalid {0} envelope")]
    InvalidEnvelope(&'static str),
}

#[derive(Deserialize)]
struct Envelope {
    v: u32,
    #[serde(rename = "type")]
    kind: String,
    #[serde(default)]
    id: Option<String>,
    #[serde(default)]
    result: Option<Value>,
    #[serde(default)]
    error: Option<RpcError>,
    #[serde(default)]
    event: Option<String>,
    #[serde(default)]
    data: Option<Value>,
}

pub fn decode_inbound(payload: &[u8]) -> Result<InboundMessage, ProtocolDecodeError> {
    if payload.len() > MAX_MESSAGE_BYTES {
        return Err(ProtocolDecodeError::MessageTooLarge);
    }
    let envelope: Envelope = serde_json::from_slice(payload)?;
    if envelope.v != PROTOCOL_VERSION {
        return Err(ProtocolDecodeError::UnsupportedVersion);
    }

    match envelope.kind.as_str() {
        "response" if envelope.result.is_some() || envelope.error.is_some() => {
            Ok(InboundMessage::Response {
                id: envelope.id,
                result: envelope.result,
                error: envelope.error,
            })
        }
        "event" => Ok(InboundMessage::Event {
            event: envelope
                .event
                .filter(|event| !event.is_empty())
                .ok_or(ProtocolDecodeError::InvalidEnvelope("event"))?,
            data: envelope.data.unwrap_or_else(|| Value::Object(Default::default())),
        }),
        "response" => Err(ProtocolDecodeError::InvalidEnvelope("response")),
        _ => Err(ProtocolDecodeError::InvalidEnvelope("message")),
    }
}
